### PR TITLE
Ensure that table exists, even when crawlers produce zero records

### LINKED ISSUE #348: Upgrade fails the Assessment Report
I installed release 0.1.1

I deleted the .ucx folder
I deleted the workflows
I deleted the dashboards
I did NOT delete the tables in the ucx database

I installed release 0.2.0
I ran the "assessment run" workflow

All of the tasks completed successfully, except for the final "assessment_report" task

The "External Locations" widget failed with:
**[TABLE_OR_VIEW_NOT_FOUND] The table or view `hive_metastore`.`ucx`.`external_locations` cannot be found. Verify the spelling and correctness of the schema and catalog. If you did not qualify the name with a schema, verify the current_schema() output, or qualify the name with the correct schema and catalog. To tolerate the error on drop use DROP VIEW IF EXISTS or DROP TABLE IF EXISTS.; line 4, pos 5**

Screenshot 2023-09-29 at 2 03 56 PM

### PR DESCRIPTION
**Pull request by FastLee**
Close #348

### CODE REVIEW

File: src/databricks/labs/ucx/framework/crawlers.py [hunk 1]
```diff
@@ -25,6 +25,10 @@ def fetch(self, sql) -> Iterator[any]:
     def save_table(self, full_name: str, rows: list[any], mode: str = "append"):
         raise NotImplementedError
 
+    @abstractmethod
```
Comment (nfx) [inline src/databricks/labs/ucx/framework/crawlers.py:28]:
```suggestion
```

Doesn't need to be abstract, implement it here

File: src/databricks/labs/ucx/framework/crawlers.py [hunk 2]
```diff
@@ -25,6 +25,10 @@ def fetch(self, sql) -> Iterator[any]:
...
+    def create_empty_table(self, full_name: str, klass):
```
Comment (nfx) [inline src/databricks/labs/ucx/framework/crawlers.py:29]:
```suggestion
    def create_table(self, full_name: str, klass):
```

File: tests/unit/framework/mocks.py [hunk 3]
```diff
@@ -14,6 +14,7 @@ def __init__(self, *, fails_on_first: dict | None = None, rows: dict | None = No
             rows = {}
         self._rows = rows
         self._save_table = []
+        self._create_table = []
```
Comment (nfx) [inline tests/unit/framework/mocks.py:17]: Why executed queries are not enough?

File: src/databricks/labs/ucx/framework/crawlers.py [hunk 4]
```diff
@@ -245,6 +256,8 @@ def _snapshot(self, fetcher, loader) -> list[any]:
 
     def _append_records(self, items):
         if len(items) == 0:
+            if self._klass is not None:
```
Comment (nfx) [inline src/databricks/labs/ucx/framework/crawlers.py:259]: These two lines could be moved to the top of the method 🤷🏻‍♂️

File: src/databricks/labs/ucx/framework/crawlers.py [hunk 5]
```diff
@@ -149,9 +155,13 @@ def save_table(self, full_name: str, rows: list[any], mode: str = "append"):
...
         df.write.saveAsTable(full_name, mode=mode)
 
+    def create_empty_table(self, full_name: str, klass):
+        ddl = f"CREATE TABLE IF NOT EXISTS {full_name} ({self._schema_for(klass)}) USING DELTA"
+        self.execute(ddl)
+
 
 class CrawlerBase:
-    def __init__(self, backend: SqlBackend, catalog: str, schema: str, table: str):
+    def __init__(self, backend: SqlBackend, catalog: str, schema: str, table: str, klass=None):
```
Comment (nfx) [inline src/databricks/labs/ucx/framework/crawlers.py:164]:
```suggestion
    def __init__(self, backend: SqlBackend, catalog: str, schema: str, table: str, klass):
```
Why klass is not required? It's needed everywhere, right?..

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/framework/crawlers.py (hunk 1 above)
Comment (FastLee) [reply src/databricks/labs/ucx/framework/crawlers.py:28]: Done

File: src/databricks/labs/ucx/framework/crawlers.py (hunk 2 above)
Comment (FastLee) [reply src/databricks/labs/ucx/framework/crawlers.py:29]: Done

File: src/databricks/labs/ucx/framework/crawlers.py (hunk 5 above)
Comment (FastLee) [reply src/databricks/labs/ucx/framework/crawlers.py:164]: I did it for legacy support. I should really not be lazy and change it everywhere...

File: src/databricks/labs/ucx/framework/crawlers.py (hunk 4 above)
Comment (FastLee) [reply src/databricks/labs/ucx/framework/crawlers.py:259]: I'm not sure what you mean.

Comment (FastLee) [reply src/databricks/labs/ucx/framework/crawlers.py:259]: Fixed

File: tests/unit/framework/mocks.py (hunk 3 above)
Comment (FastLee) [reply tests/unit/framework/mocks.py:17]: Fixed

File: src/databricks/labs/ucx/framework/crawlers.py [hunk 6]
```diff
@@ -82,18 +86,14 @@ def fetch(self, sql) -> Iterator[any]:
...
+    def save_table(self, full_name: str, rows: list[any], klass: type, mode="append"):
         if mode == "overwrite":
             msg = "Overwrite mode is not yet supported"
             raise NotImplementedError(msg)
         rows = self._filter_none_rows(rows, full_name)
+        self.create_table(full_name, klass)
         if len(rows) == 0:
             return
-
-        klass = rows[0].__class__
```
Comment (nfx) [inline src/databricks/labs/ucx/framework/crawlers.py:93]: Previously we got the klass by looking at the first record, now we add it as the explicit argument. Do we really want to add explicit argument?

**Review by nfx — APPROVED** [review]

**Pull request merged**